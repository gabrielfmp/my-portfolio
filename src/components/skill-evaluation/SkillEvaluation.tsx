import { Dot, SkillEvaluationWrapper, SkillText } from './styles';
import HalfDot from './HalfDot';
import React, { FunctionComponent } from 'react';

type Props = {
  skill: string;
  evaluation: number;
  outOf: number;
};

const SkillEvaluation: FunctionComponent<Props> = ({
  skill,
  evaluation,
  outOf,
}) => {
  // To know the 0.5 of 4.5
  const fractionalPart = evaluation % 1;

  // For simplicity, since I don't want to have 1.312 dots filled, when
  // evaluation has a fractional part this components renders half dot
  const hasHalfDot = !!fractionalPart;

  // Make sure filledDotsNumber is integer
  const filledDotsNumber = evaluation - fractionalPart;

  // When we have an half dots, create one less grey dot
  const remainingDots = hasHalfDot
    ? outOf - filledDotsNumber - 1
    : outOf - filledDotsNumber;

  return (
    <SkillEvaluationWrapper>
      <SkillText>{skill}</SkillText>

      {
        // Making *filledDotsNumber* black dots
        Array(filledDotsNumber)
          .fill(undefined)
          .map((_, i) => (
            <Dot color="#000" key={i} data-testid="filled-dot" />
          ))
      }

      {
        // Making one half black half grey dots
        hasHalfDot && <HalfDot />
      }

      {
        // Making *remainingDots* grey dots
        Array(remainingDots)
          .fill(undefined)
          .map((_, i) => (
            <Dot color="#BCBCBC" key={i} data-testid="empty-dot" />
          ))
      }
    </SkillEvaluationWrapper>
  );
};

export default SkillEvaluation;
