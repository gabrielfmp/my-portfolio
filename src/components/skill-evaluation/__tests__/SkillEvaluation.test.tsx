import { render } from '@testing-library/react';
import React from 'react';
import SkillEvaluation from '../SkillEvaluation';

test('snapshot', () => {
  const { getAllByTestId, queryByTestId, asFragment } = render(
    <SkillEvaluation skill="foo" evaluation={1} outOf={2} />
  );

  expect(getAllByTestId('filled-dot').length).toBe(1);
  expect(queryByTestId('half-dot')).toBeNull();
  expect(getAllByTestId('empty-dot').length).toBe(1);

  expect(asFragment()).toMatchSnapshot();
});

test('rendering half dots', () => {
  const { getAllByTestId } = render(
    <SkillEvaluation skill="foo" evaluation={5.5} outOf={7} />
  );

  expect(getAllByTestId('filled-dot').length).toBe(5);
  expect(getAllByTestId('half-dot').length).toBe(1);
  expect(getAllByTestId('empty-dot').length).toBe(1);
});
