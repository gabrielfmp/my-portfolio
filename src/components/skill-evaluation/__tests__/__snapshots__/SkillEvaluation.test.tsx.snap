// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1tv79r0"
  >
    <p
      class="css-1x26r7c"
    >
      foo
    </p>
    <div
      class="css-1dpzy13"
      color="#000"
      data-testid="filled-dot"
    />
    <div
      class="css-5uorsv"
      color="#BCBCBC"
      data-testid="empty-dot"
    />
  </div>
</DocumentFragment>
`;
